### PR TITLE
ASCS Source Implementation

### PR DESCRIPTION
**Pull request by zxzxwu**
* Implement ASE state transition for source
* Make ASE emit state changed events
* Replace pdu event with sink which is more reasonable and have a better type annotation
* Implement a LEA app
* Improve CIS handling in device and host

### CODE REVIEW

**Comment by zxzxwu** [comment]
SCO has the same issue. I think one solution could be - again, using microphone input from a web frontend.

**Comment by barbibulle** [comment]
> SCO has the same issue. I think one solution could be - again, using microphone input from a web frontend.

You could read the input from an LC3 encoded audio file, like for the A2DP source example.

**Comment by zxzxwu** [comment]
Consider the difficulty of audio prcessing and state synchronization between Python and Web Frontend, I think it's better to host codec in Python instead of JS.

The liblc3.wasm file is built from liblc3 source. With wasmtime, it could be executed on all platforms (with CPython only?). Not sure whether there is a better way to pack liblc3?

**Comment by zxzxwu** [comment]
Not sure how to solve this...
```
thread 'tokio-runtime-worker' panicked at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-asyncio-0.18.0/src/testing.rs:285:39:
called `Result::unwrap()` on an `Err` value: PyErr { type: <class 'ModuleNotFoundError'>, value: ModuleNotFoundError("No module named 'cryptography'"), traceback: Some(<traceback object at 0x7bb4710b75c0>) }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
pyo3_asyncio.RustPanic: rust future panicked
thread 'main' panicked at pytests/pytests.rs:15:1:
called `Result::unwrap()` on an `Err` value: ()
error: test failed, to rerun pass `--test pytests`
```

File: setup.cfg [hunk 1]
```diff
@@ -96,6 +96,7 @@ development =
     types-appdirs >= 1.4.3
     types-invoke >= 1.7.3
     types-protobuf >= 4.21.0
+    wasmtime == 1.0.0
```
Comment (barbibulle) [inline setup.cfg:99]: It seems that the current version of `wasmtime` is `20.0.0`. Is `1.0.0` really what's needed here?

**Comment by barbibulle** [comment]
> Not sure how to solve this...
> 
> ```
> thread 'tokio-runtime-worker' panicked at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-asyncio-0.18.0/src/testing.rs:285:39:
> called `Result::unwrap()` on an `Err` value: PyErr { type: <class 'ModuleNotFoundError'>, value: ModuleNotFoundError("No module named 'cryptography'"), traceback: Some(<traceback object at 0x7bb4710b75c0>) }
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> Traceback (most recent call last):
>   File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
>     return future.result()
> pyo3_asyncio.RustPanic: rust future panicked
> thread 'main' panicked at pytests/pytests.rs:15:1:
> called `Result::unwrap()` on an `Err` value: ()
> error: test failed, to rerun pass `--test pytests`
> ```

When do you get this error?

File: setup.cfg (hunk 1 above)
Comment (zxzxwu) [reply setup.cfg]: Somehow I always get 1.0.0 when searching on Google 🤔 Updated.

**Comment by zxzxwu** [comment]
Since we started working on Broadcast, we may merge this PR first to avoid conflict.

**Comment by barbibulle** [comment]
> Since we started working on Broadcast, we may merge this PR first to avoid conflict.

Yes. I wasn't sure if you were ready with this one yet or not. I'll do a quick re-review today.

**Comment by zxzxwu** [comment]
Yes, I think it's ready.

File: apps/lea_unicast/app.py [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+                        (
+                            AdvertisingData.COMPLETE_LOCAL_NAME,
+                            bytes(device_config.name, 'utf-8'),
+                        ),
+                        (
+                            AdvertisingData.FLAGS,
+                            bytes(
+                                [
+                                    AdvertisingData.LE_GENERAL_DISCOVERABLE_MODE_FLAG
+                                    | AdvertisingData.BR_EDR_HOST_FLAG
```
[the remark this answers is not in the document]
Comment (zxzxwu) [reply apps/lea_unicast/app.py:481]: Nope. Done.

File: bumble/profiles/bap.py [hunk 3]
```diff
@@ -78,6 +78,10 @@ class AudioLocation(enum.IntFlag):
     LEFT_SURROUND           = 0x04000000
     RIGHT_SURROUND          = 0x08000000
 
+    @property
+    def num_channels(self) -> int:
```
[the remark this answers is not in the document]
Comment (zxzxwu) [reply bumble/profiles/bap.py:82]: Done

File: bumble/profiles/bap.py [hunk 4]
```diff
@@ -1044,8 +1074,9 @@ def state(self) -> State:
 
     @state.setter
     def state(self, new_state: State) -> None:
-        logger.debug(f'{self} state change -> {colors.color(new_state.name, "cyan")}')
+        logger.info(f'{self} state change -> {colors.color(new_state.name, "cyan")}')
```
[the remark this answers is not in the document]
Comment (zxzxwu) [reply bumble/profiles/bap.py:1077]: Done

File: bumble/profiles/bap.py [hunk 5]
```diff
@@ -1044,8 +1074,9 @@ def state(self) -> State:
...
         self._state = new_state
+        self.emit('state_change', new_state)
```
[the remark this answers is not in the document]
Comment (zxzxwu) [reply bumble/profiles/bap.py:1079]: Sounds reasonable. It is really a pain point to use eventemitter - Despite it benefits Bumble a lot, the missing of type hint is something unstable.

File: examples/run_unicast_server.py [hunk 6]
```diff
@@ -45,6 +53,32 @@
...
+                | SupportedSamplingFrequency.FREQ_16000
+                | SupportedSamplingFrequency.FREQ_24000
+                | SupportedSamplingFrequency.FREQ_32000
+                | SupportedSamplingFrequency.FREQ_48000
+            ),
+            supported_frame_durations=(
+                SupportedFrameDuration.DURATION_7500_US_SUPPORTED
+                | SupportedFrameDuration.DURATION_10000_US_SUPPORTED
+            ),
+            supported_audio_channel_counts=[1, 2],
```
[the remark this answers is not in the document]
Comment (zxzxwu) [reply examples/run_unicast_server.py:71]: Done

**Pull request merged**